File: LinkedList/206 Reverse Linked List/approaches/iterative/iterative.py
# Problem: 206. Reverse Linked List
# Approach: Iterative
# Language: python3

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        curr=head
        prev=None
        while curr!=None:
            nextele=curr.next
            curr.next=prev
            prev=curr
            curr=nextele
        return prev
        

File: LinkedList/Reverse Linked List/approaches/iterative/iterative.cpp
# Problem: Reverse Linked List
# Approach: Iterative
# Language: cpp
# Time: O(n)
# Space: O(1)

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* reverseList(ListNode* head) {
        ListNode* curr=head;
        ListNode* prev = nullptr;
        ListNode* newnode=head;
        while (curr!=nullptr){
            newnode=curr->next;
            curr->next=prev;
            prev=curr;
            curr=newnode;
        }
        return prev;
    }
};